packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/macro_expand.rs

```diff
@@ -0,0 +1,255 @@
+/// a `<line />` maps to a `<line extend="..." />`, etc.)
+pub struct Expander {}
+
+impl Expander {
+    /// Expand all macros and function macros into their "xml" form. After expansion,
+    /// the resulting tree may not be serializable as XML since element attributes may contain
+    /// other elements.
+    pub fn expand(flat_root: &mut FlatRoot) {
+        let resolver = Resolver::from_flat_root(flat_root);
+        Expander::expand_macros(flat_root, &resolver);
+        Expander::consume_extend_attributes(flat_root);
+    }
+
+    /// Expand all macros and function macros into their "xml" form.
+    fn expand_macros(flat_root: &mut FlatRoot, resolver: &Resolver) {
+        for idx in 0..flat_root.nodes.len() {
+            // The original `nodes[idx]` node is being completely replaced, so we are free to take its value,
+            // which will prevent the borrow checker from complaining if we mutate `flat_root` during processing.
+            flat_root.nodes[idx] = match mem::take(&mut flat_root.nodes[idx]) {
+                FlatNode::Macro(macro_) => {
+                    // A macro $f get's replace with <foo extend="f" /> where `foo` is the tag name of the referent.
+                    // e.g. <point name="p" />$p becomes <point name="p" /><point extend="p" />
+                    // Expanding a macro is can be done with a single replacement.
+
+                    //flat_root.nodes[idx] = resolved
+                    match resolver.resolve(&macro_.path, macro_.idx) {
+                        Ok(ref_resolution) => {
+                            // Get the tag name of the referent
+                            let name = match &flat_root.nodes[ref_resolution.node_idx] {
+                                FlatNode::Element(e) => e.name.clone(),
+                                _ => panic!("Expected an element"),
+                            };
+                            FlatNode::Element(FlatElement {
+                                idx: macro_.idx,
+                                parent: macro_.parent,
+                                attributes: Vec::new(),
+                                children: Vec::new(),
+                                name,
+                                position: macro_.position.clone(),
+                                extending: Some(ref_resolution),
+                            })
+                        }
+                        Err(err) => FlatNode::Error(FlatError {
+                            idx: macro_.idx,
+                            parent: macro_.parent,
+                            message: format!("Macro resolution error: {}", err),
+                            position: macro_.position.clone(),
+                        }),
+                    }
+                }
+                FlatNode::FunctionMacro(function_macro) => {
+                    // A function ref `$$f(x)` becomes `<evaluate extend="f"><ol><li>x</li></ol></evaluate>`
```

Comment:
Currently, the evaluate syntax is 
`<evaluate function="$f" input="x" />`
or
`<evaluate function="$f" input="x y" />`
where the `input` attribute gets expanded into a list. I don't know to what extent people use the `<evaluate>` tag, but if they did, I think it would be unintuitive to put the lists as children.

It's true that one cannot enter a list directly into an attribute, but we have sugar that translates a space separated list implemented on the list component (at least on the `<mathList>` component, not currently on the `<li>` component).